File: server/src/services/hwAccel.ts
/**
 * hwAccel — runtime detection of hardware-accelerated video encode backends.
 *
 * The server probes at startup for a platform-appropriate HW path. Success
 * caches a tagged-union config that the ffmpeg command builder and chunker
 * read synchronously.
 *
 * Policy: if HW_ACCEL=auto (the default) and the probe fails, the server
 * exits with a clear diagnostic — NOT silently falling back to software.
 * Software 4K encode has been measured at ~0.8× realtime on modern desktop
 * CPUs and produces continuous playback stalls; shipping that by default
 * would mask real infrastructure regressions. Users who deliberately want
 * software mode set HW_ACCEL=off; per-chunk fallback still uses software
 * when a transient HW error hits mid-session.
 *
 * Today only the VAAPI branch is fully implemented; macOS (VideoToolbox)
 * and Windows (QSV / NVENC / AMF) variants are present in the type system
 * and throw on probe so the Rust/Tauri port has an obvious checklist.
 */
import { spawnSync } from "node:child_process";

import { getOtelLogger } from "../telemetry/index.js";

const log = getOtelLogger("hwAccel");

export type HwAccelKind = "software" | "vaapi" | "videotoolbox" | "qsv" | "nvenc" | "amf";

export type HwAccelConfig =
  | { kind: "software" }
  | { kind: "vaapi"; device: string }
  | { kind: "videotoolbox" }
  | { kind: "qsv" }
  | { kind: "nvenc" }
  | { kind: "amf" };

const DEFAULT_VAAPI_DEVICE = "/dev/dri/renderD128";

let detected: HwAccelConfig | null = null;

/**
 * Returns the memoised HW accel config decided at startup. Throws if called
 * before `detectHwAccel()` — the server's startup sequence must probe first.
 */
export function getHwAccelConfig(): HwAccelConfig {
  if (!detected) {
    throw new Error(
      "getHwAccelConfig() called before detectHwAccel(). The server must probe HW accel during startup."
    );
  }
  return detected;
}

/**
 * Probe the environment and cache the resulting config. Called once at
 * server startup from `index.ts`. Exits the process on fatal HW misconfig
 * (see module doc for rationale).
 */
export async function detectHwAccel(
  ffmpegPath: string,
  mode: "auto" | "off"
): Promise<HwAccelConfig> {
  if (detected) return detected;

  if (mode === "off") {
    log.info("Hardware acceleration explicitly disabled (HW_ACCEL=off) — software encode", {
      mode,
    });
    detected = { kind: "software" };
    return detected;
  }

  if (process.platform === "linux") {
    const ok = probeVaapi(ffmpegPath, DEFAULT_VAAPI_DEVICE);
    if (ok.ok) {
      detected = { kind: "vaapi", device: DEFAULT_VAAPI_DEVICE };
      log.info(`Hardware acceleration selected — vaapi (${DEFAULT_VAAPI_DEVICE})`, {
        kind: "vaapi",
        device: DEFAULT_VAAPI_DEVICE,
      });
      return detected;
    }
    fatal(
      `Hardware acceleration probe failed on linux.\n` +
        `  device: ${DEFAULT_VAAPI_DEVICE}\n` +
        `  exit:   ${ok.exit}\n` +
        `  stderr: ${ok.stderr.slice(0, 800)}\n\n` +
        `Common causes:\n` +
        `  • GPU driver too old for your hardware (Lunar Lake needs intel-media-driver 24.2.0+)\n` +
        `  • User not in 'render' group or ACL missing on ${DEFAULT_VAAPI_DEVICE}\n` +
        `  • Running inside a container/VM without GPU passthrough\n\n` +
        `Set HW_ACCEL=off to force software mode (note: software 4K encode stalls continuously).`
    );
  }

  if (process.platform === "darwin") {
    fatal(
      `HW accel on darwin not yet implemented.\n` +
        `Set HW_ACCEL=off to run software encode, or contribute a videotoolbox implementation ` +
        `in server/src/services/hwAccel.ts + the 'videotoolbox' branch in ffmpegFile.applyOutputOptions.`
    );
  }

  if (process.platform === "win32") {
    fatal(
      `HW accel on win32 not yet implemented.\n` +
        `Set HW_ACCEL=off to run software encode, or contribute a qsv/nvenc/amf implementation ` +
        `in server/src/services/hwAccel.ts + the matching branch in ffmpegFile.applyOutputOptions.`
    );
  }

  fatal(`Unsupported platform: ${process.platform}. Set HW_ACCEL=off to force software mode.`);
}

/**
 * VAAPI smoke test. Runs a 0.1 s synthetic encode through h264_vaapi; exit
 * code 0 means the device initialised and the encoder produced frames.
 */
function probeVaapi(
  ffmpegPath: string,
  device: string
): { ok: true } | { ok: false; exit: number | null; stderr: string } {
  const args = [
    "-hide_banner",
    "-v",
    "error",
    "-init_hw_device",
    `vaapi=va:${device}`,
    "-f",
    "lavfi",
    "-i",
    "testsrc=duration=0.1:size=320x240:rate=24",
    "-vf",
    "format=nv12,hwupload",
    "-c:v",
    "h264_vaapi",
    "-qp",
    "23",
    "-f",
    "null",
    "-",
  ];
  const result = spawnSync(ffmpegPath, args, { encoding: "utf8", timeout: 10_000 });
  if (result.status === 0) return { ok: true };
  return { ok: false, exit: result.status, stderr: result.stderr || result.stdout || "" };
}

/** Log a fatal error and exit the process. The OTel log call is best-effort
 *  for Seq; the plain stderr write guarantees the user sees it in the
 *  terminal even if OTel export hasn't flushed yet. */
function fatal(message: string): never {
  log.error(`Hardware acceleration startup failed:\n${message}`, {});
  process.stderr.write(`\n[xstream] Hardware acceleration startup failed:\n${message}\n\n`);
  process.exit(1);
}

File: server-rust/src/services/hw_accel.rs
//! Hardware-acceleration probe + selection.
//!
//! Policy: the server probes at startup for a platform-appropriate HW path.
//! If `auto` (the default) and the probe fails, the resolver returns a typed
//! error — `main()` surfaces it as `AppError::HwAccelProbe` and the process
//! exits with a clear diagnostic. Silent fallback to software is deliberately
//! NOT the default: 4K libx264 stalls continuously on consumer CPUs and would
//! mask real driver / permission regressions in production.
//!
//! Today only the VAAPI branch is fully implemented. macOS (VideoToolbox)
//! and Windows (QSV / NVENC / AMF) variants are tracked as
//! `Plan/Open-Questions.md §1` follow-ups; this module returns
//! `HwAccelError::PlatformNotImplemented` for those hosts so the failure
//! surface is explicit rather than silent.

use std::path::Path;
use std::time::Duration;

use thiserror::Error;
use tokio::process::Command;
use tokio::time::timeout;

use crate::services::ffmpeg_file::HwAccelConfig;

/// Mode selector — usually read from the `HW_ACCEL` env var
/// (`auto` is the implicit default; only `off` is meaningful as an opt-out).
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum HwAccelMode {
    Auto,
    Off,
}

impl HwAccelMode {
    /// `HW_ACCEL=off` → `Off`; anything else (including unset, `auto`,
    /// or any other value) → `Auto`. The default is opt-in HW accel; only
    /// the explicit `off` opt-out is honored.
    pub fn from_env() -> Self {
        match std::env::var("HW_ACCEL").as_deref() {
            Ok("off") => Self::Off,
            _ => Self::Auto,
        }
    }
}

pub const DEFAULT_VAAPI_DEVICE: &str = "/dev/dri/renderD128";

#[derive(Debug, Error)]
pub enum HwAccelError {
    #[error("spawning ffmpeg at {ffmpeg} for VAAPI probe: {source}")]
    ProbeSpawn {
        ffmpeg: std::path::PathBuf,
        #[source]
        source: std::io::Error,
    },

    #[error("ffmpeg VAAPI probe timed out after {timeout_ms} ms — driver / permissions issue?")]
    ProbeTimeout { timeout_ms: u64 },

    #[error(
        "VAAPI probe failed on linux\n  device: {device}\n  exit:   {exit:?}\n  stderr: {stderr}\n\n\
         Common causes:\n\
           • GPU driver too old for your hardware (Lunar Lake needs intel-media-driver 24.2.0+)\n\
           • User not in 'render' group or ACL missing on {device}\n\
           • Running inside a container/VM without GPU passthrough\n\n\
         Set HW_ACCEL=off to force software mode (note: software 4K encode stalls continuously)."
    )]
    VaapiProbeFailed {
        device: String,
        exit: Option<i32>,
        stderr: String,
    },

    #[error(
        "HW accel on {os} not yet implemented.\n\
         Set HW_ACCEL=off to run software encode, or contribute the {hint} \
         implementation in server-rust/src/services/hw_accel.rs and the matching \
         branch in services/ffmpeg_file.rs::build_encode_argv."
    )]
    PlatformNotImplemented {
        os: &'static str,
        hint: &'static str,
    },
}

pub type HwAccelResult<T> = Result<T, HwAccelError>;

/// Resolve the HW accel config for this host. Called once at startup.
///
/// `mode = Off` always returns `Software` immediately (no probe, no I/O).
/// `mode = Auto` runs a 0.1 s synthetic encode through `h264_vaapi` on
/// Linux; success → `Vaapi { device }`, failure → typed error so `main()`
/// can decide policy (today: fatal exit; future Tauri build per
/// `Plan/Open-Questions.md §4` may degrade to `Software` with a toast).
pub async fn resolve_hw_accel(ffmpeg: &Path, mode: HwAccelMode) -> HwAccelResult<HwAccelConfig> {
    if mode == HwAccelMode::Off {
        return Ok(HwAccelConfig::Software);
    }

    if cfg!(target_os = "linux") {
        probe_vaapi(ffmpeg, DEFAULT_VAAPI_DEVICE).await?;
        return Ok(HwAccelConfig::Vaapi {
            device: DEFAULT_VAAPI_DEVICE.to_string(),
        });
    }

    if cfg!(target_os = "macos") {
        return Err(HwAccelError::PlatformNotImplemented {
            os: "darwin",
            hint: "videotoolbox",
        });
    }

    if cfg!(target_os = "windows") {
        return Err(HwAccelError::PlatformNotImplemented {
            os: "win32",
            hint: "qsv/nvenc/amf",
        });
    }

    Err(HwAccelError::PlatformNotImplemented {
        os: "unknown",
        hint: "none",
    })
}

const PROBE_TIMEOUT_MS: u64 = 10_000;

/// Synthetic VAAPI encode — `lavfi:testsrc` → `h264_vaapi` for 0.1 s. Exit 0
/// means the device initialised, the encoder produced frames, and ffmpeg
/// closed cleanly. Anything else surfaces as `VaapiProbeFailed` with the
/// stderr tail for diagnosis.
async fn probe_vaapi(ffmpeg: &Path, device: &str) -> HwAccelResult<()> {
    let args = ["-hide_banner", "-v", "error", "-init_hw_device"];
    let device_arg = format!("vaapi=va:{device}");
    let probe_args: Vec<&str> = args
        .iter()
        .copied()
        .chain([
            device_arg.as_str(),
            "-f",
            "lavfi",
            "-i",
            "testsrc=duration=0.1:size=320x240:rate=24",
            "-vf",
            "format=nv12,hwupload",
            "-c:v",
            "h264_vaapi",
            "-qp",
            "23",
            "-f",
            "null",
            "-",
        ])
        .collect();

    let fut = Command::new(ffmpeg).args(&probe_args).output();
    let output = timeout(Duration::from_millis(PROBE_TIMEOUT_MS), fut)
        .await
        .map_err(|_| HwAccelError::ProbeTimeout {
            timeout_ms: PROBE_TIMEOUT_MS,
        })?
        .map_err(|source| HwAccelError::ProbeSpawn {
            ffmpeg: ffmpeg.to_path_buf(),
            source,
        })?;

    if output.status.success() {
        return Ok(());
    }

    Err(HwAccelError::VaapiProbeFailed {
        device: device.to_string(),
        exit: output.status.code(),
        stderr: String::from_utf8_lossy(&output.stderr)
            .chars()
            .take(800)
            .collect(),
    })
}

// ── Tests ────────────────────────────────────────────────────────────────────
//
// The full VAAPI probe needs a real ffmpeg binary AND a real `/dev/dri`
// device, so it can't run in CI. We cover the deterministic surface:
// mode parsing, the early-Software path, and the not-implemented branches
// for non-Linux hosts. Real-binary VAAPI parity is asserted in the
// integration tests for chunker/encode (gated on XSTREAM_TEST_MEDIA_DIR).

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mode_off_short_circuits_to_software() {
        // Use a non-existent ffmpeg path — Off should not spawn anything,
        // so the path isn't dereferenced.
        let dummy = Path::new("/does/not/exist/ffmpeg");
        let rt = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .expect("runtime");
        let cfg = rt
            .block_on(resolve_hw_accel(dummy, HwAccelMode::Off))
            .expect("Off must short-circuit");
        assert_eq!(cfg, HwAccelConfig::Software);
    }

    #[test]
    fn auto_on_linux_with_missing_ffmpeg_returns_probe_spawn_error() {
        if !cfg!(target_os = "linux") {
            return;
        }
        let dummy = Path::new("/definitely/not/a/real/binary/ffmpeg");
        let rt = tokio::runtime::Builder::new_current_thread()
            .enable_all()
            .build()
            .expect("runtime");
        let err = rt
            .block_on(resolve_hw_accel(dummy, HwAccelMode::Auto))
            .expect_err("auto with bogus ffmpeg must fail");
        assert!(
            matches!(err, HwAccelError::ProbeSpawn { .. }),
            "expected ProbeSpawn, got {err:?}"
        );
    }

    #[test]
    fn from_env_treats_off_literally() {
        // Exercise the literal-value branches without depending on the
        // ambient HW_ACCEL — clear it for the duration of the assertions.
        let prior = std::env::var("HW_ACCEL").ok();
        // SAFETY: tests in this module run sequentially via the test harness,
        // and HwAccelMode::from_env reads the env var fresh on each call.
        // We restore the original value after the assertions.
        unsafe { std::env::remove_var("HW_ACCEL") };
        assert_eq!(HwAccelMode::from_env(), HwAccelMode::Auto);
        unsafe { std::env::set_var("HW_ACCEL", "off") };
        assert_eq!(HwAccelMode::from_env(), HwAccelMode::Off);
        unsafe { std::env::set_var("HW_ACCEL", "auto") };
        assert_eq!(HwAccelMode::from_env(), HwAccelMode::Auto);
        unsafe { std::env::set_var("HW_ACCEL", "garbage") };
        // Anything that isn't literally "off" falls back to Auto — only
        // the explicit opt-out is honored.
        assert_eq!(HwAccelMode::from_env(), HwAccelMode::Auto);
        match prior {
            Some(v) => unsafe { std::env::set_var("HW_ACCEL", v) },
            None => unsafe { std::env::remove_var("HW_ACCEL") },
        }
    }
}
